src/Channels/UnownedBufferChannel.cs

```diff
@@ -0,0 +1,352 @@
+        {
+            return WriteAsync(new UnownedBuffer(buffer), cancellationToken);
+        }
+
+        /// <summary>
+        /// Writes a new buffer into the channel. The task returned by this operation only completes when the next
+        /// Read has been queued, or the Reader has completed, since the buffer provided here needs to be kept alive
+        /// until the matching Read finishes (because we don't have ownership tracking when working with unowned buffers)
+        /// </summary>
+        /// <param name="buffer"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        // Called by the WRITER
+        public async Task WriteAsync(IBuffer buffer, CancellationToken cancellationToken)
+        {
+            // If Writing has stopped, why is the caller writing??
+            if (Writing.Status != TaskStatus.WaitingForActivation)
+            {
+                throw new OperationCanceledException("Writing has ceased on this Channel");
+            }
+
+            // If Reading has stopped, we cancel. We don't write unless there's a reader ready in this channel.
+            if (Reading.Status != TaskStatus.WaitingForActivation)
+            {
+                throw new OperationCanceledException("Reading has ceased on this Channel");
+            }
+
+            // Register for cancellation on this token for the duration of the write
+            using (cancellationToken.Register(state => ((UnownedBufferChannel)state).CancelWriter(), this))
+            {
+                // Wait for reading to start
+                await ReadingStarted;
+
+                // Cancel this task if this write is cancelled
+                cancellationToken.ThrowIfCancellationRequested();
+
+                // Allocate a new segment to hold the buffer being written.
+                var segment = new BufferSegment(buffer);
+                segment.End = buffer.Data.Length;
+
+                if (_head == null)
+                {
+                    // Update the head to point to the head of the buffer.
+                    _head = segment;
+                }
+                else if (_tail != null)
+                {
+                    // Add this segment to the end of the chain
+                    _tail.Next = segment;
+                }
+
+                // Always update tail to the buffer's tail
+                _tail = segment;
+
+                // Trigger the continuation
+                Complete();
+
+                // Wait for another read to come (or for the end of Reading, which will also trigger this gate to open) in before returning
+                await _readWaiting;
+
+                // We need to preserve any buffers that haven't been consumed
+                var current = _head;
+                while (current != null)
+                {
+                    current.Buffer = current.Buffer.Preserve(0, current.Buffer.Data.Length);
```

Comment:
Wouldn't it be more efficient to call `current.Buffer.Preserve(current.Start, current.Start - current.End)`? The only thing you would need to consider is if `current.Start` and `current.End` need to be updated since the underlying `current.Buffer` will sometimes change (in the unowned case), but not always (in the owned case).